src/System.Text.Primitives/System/Text/Encoders/UnicodeScalar.cs

```diff
@@ -0,0 +1,246 @@
+        /// </summary>
+        public static explicit operator int(UnicodeScalar value) => value.Value;
+
+        public static bool operator ==(UnicodeScalar a, UnicodeScalar b) => (a.Value == b.Value);
+
+        public static bool operator !=(UnicodeScalar a, UnicodeScalar b) => (a.Value != b.Value);
+
+        public static bool operator <(UnicodeScalar a, UnicodeScalar b) => (a.Value < b.Value);
+
+        public static bool operator <=(UnicodeScalar a, UnicodeScalar b) => (a.Value <= b.Value);
+
+        public static bool operator >(UnicodeScalar a, UnicodeScalar b) => (a.Value > b.Value);
+
+        public static bool operator >=(UnicodeScalar a, UnicodeScalar b) => (a.Value >= b.Value);
+
+        private bool IsValid => IsValidScalar((uint)Value);
+
+        /// <summary>
+        /// Returns the number of UTF-16 code units (<see cref="char"/>s) required to represent this scalar.
+        /// Scalars in the range [U+0000..U+FFFF] require one UTF-16 code unit.
+        /// Scalars in the range [U+10000..U+10FFFF] require two UTF-16 code units.
+        /// </summary>
+        public int Utf16CodeUnitCount
+        {
+            get
+            {
+                Debug.Assert(IsValid);
+                return ((uint)Value < 0x10000U) ? 1 : 2;
+            }
+        }
+
+        /// <summary>
+        /// Returns the number of UTF-8 code units (<see cref="byte"/>s) required to represent this scalar.
+        /// Scalars in the range [U+0000..U+007F] require one UTF-8 code unit.
+        /// Scalars in the range [U+0080..U+07FF] require two UTF-8 code units.
+        /// Scalars in the range [U+0800..U+FFFF] require three UTF-8 code unit.
+        /// Scalars in the range [U+10000..U+10FFFF] require four UTF-8 code units.
+        /// </summary>
+        public int Utf8CodeUnitCount
+        {
+            get
+            {
+                Debug.Assert(IsValid);
+
+                if ((uint)Value < 0x80U)
+                {
+                    return 1;
+                }
+                else if ((uint)Value < 0x800U)
+                {
+                    return 2;
+                }
+                else if ((uint)Value < 0x10000U)
+                {
+                    return 3;
+                }
+                else
+                {
+                    return 4;
+                }
+            }
+        }
+
+        public int CompareTo(UnicodeScalar other)
+        {
+            return this.Value.CompareTo(other.Value);
+        }
+
+        /// <summary>
+        /// Copies the UTF-16 code unit representation of this scalar to an output buffer.
+        /// The buffer must be large enough to hold the required number of <see cref="char"/>s.
+        /// The <see cref="Utf16CodeUnitCount"/> property gives the required output buffer length.
+        /// </summary>
+        public void CopyUtf16CodeUnitsTo(Span<char> utf16)
```

Comment:
Wouldn't it be useful if the routine returned the number of code units written?

Reply:
There’s a property that gives you that information and the caller needs to query the property ahead of time, so it seems redundant. Is your suggestion a TryCopy method?

Reply:
I see. Thanks.